agent/engine/daemonmanager/daemon_manager_linux.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//      http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package daemonmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"os"
+
+	"github.com/pborman/uuid"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apicontainerstatus "github.com/aws/amazon-ecs-agent/agent/api/container/status"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/amazon-ecs-agent/agent/utils/loader"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	md "github.com/aws/amazon-ecs-agent/ecs-agent/manageddaemon"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/docker/docker/api/types"
+	dockercontainer "github.com/docker/docker/api/types/container"
+)
+
+const (
+	// all daemons will share the same user id
+	// note: AppNetUID is 20000
+	daemonUID                       = 20001
+	ecsAgentLogFileENV              = "ECS_LOGFILE"
+	defaultECSAgentLogPathContainer = "/log"
+)
+
+// each daemon manager manages one single daemon container
+type daemonManager struct {
+	managedDaemon *md.ManagedDaemon
+}
+
+func NewDaemonManager(manageddaemon *md.ManagedDaemon) DaemonManager {
+	return &daemonManager{managedDaemon: manageddaemon}
+}
+
+func (dm *daemonManager) CreateDaemonTask() (*apitask.Task, error) {
+	imageName := dm.managedDaemon.GetLoadedDaemonImageRef()
+	containerRunning := apicontainerstatus.ContainerRunning
+	dockerHostConfig := dockercontainer.HostConfig{
+		NetworkMode: apitask.HostNetworkMode,
+		// the default value of 0 for MaximumRetryCount means retry indefinitely
+		RestartPolicy: dockercontainer.RestartPolicy{
+			Name:              "on-failure",
+			MaximumRetryCount: 0,
+		},
+	}
+	if !dm.managedDaemon.IsValidManagedDaemon() {
+		return nil, fmt.Errorf("%s is an invalid managed daemon", dm.managedDaemon.GetImageName())
```

Comment:
nit: dm.managedDaemon.GetImageName() could be defined in one place and reused here and on L99